### PR TITLE
evm: use `RethEvmBuilder` for `Optimism` config

### PR DESCRIPTION
**Pull request by tcoratger**
As a follow up to our use of the Evm builder (related https://github.com/paradigmxyz/reth/pull/10210), this PR uses the `RethEvmBuilder` for the Optimism configuration in order to gradually eliminate the `EvmBuilder` from the codebase.
- The unit tests added in https://github.com/paradigmxyz/reth/pull/10280 should cover the entire EVM building for the Optimism cfg. I hope I didn't miss any untested edge cases that would be problematic.
- Before we used the `optimism()` function which is part of the `EvmBuilder` https://github.com/bluealloy/revm/blob/1ad860469755e3cf71383f45d71c3faaf61d3029/crates/revm/src/builder.rs#L162-L169 to modify the handler to apply the spec ID changes needed by Optimism but with the new approach using `RethEvmBuilder`, it is not practical to do like that. So I opted for another approach which consists of first building a default EVM which is an Ethereum EVM before modifying the handler.

@mattsse Don't hesitate to tell me what you think and if the chosen approach seems correct to you. The goal of this PR is really to extend the usage of our `RethEvmBuilder` and to no longer have an `EvmBuilder` in the Optimism implementation.

### CODE REVIEW

File: crates/optimism/evm/src/lib.rs [hunk 1]
```diff
@@ -114,20 +115,33 @@ impl ConfigureEvm for OptimismEvmConfig {
...
-            .build()
+        // Define a default EVM configuration with an inspector
+        let mut evm = {
+            self.default_external_context();
+            RethEvmBuilder::new(db, ())
+        }
+        .build_with_inspector(inspector);
+
+        // Apply the Optimism configuration to the EVM
+        evm.handler = Handler::optimism_with_spec(evm.handler.cfg.spec_id);
```
Comment (mattsse) [inline crates/optimism/evm/src/lib.rs:143]: is this correct?

isn't the inspector installed as a callback in the handler, so if we overwrite it like this, we lose it?

Comment (mattsse) [reply crates/optimism/evm/src/lib.rs:143]: > My idea here was to do everything to completely remove the REVM builder from the codebase.

ideally, yes but I really don't know if this is correct because from my understanding this removes the inspector

**Comment by tcoratger** [comment]
Closing for now as it probably needs more work to remove optimism feature flag in order to handle this in an easier way.
